### PR TITLE
Contributing guidelines

### PR DESCRIPTION
**Pull request by LongLiveCHIEF**
@jazzpi @lloeki I added Contributing guidelines, and a pull request template. I'd like a `+1` from both of you on this one before I make it official.

### CODE REVIEW

**Comment by jazzpi** [comment]
Looks mostly good, but you added PULL_REQUEST_TEMPLATE.MD and PULL_REQUEST_TEMPLATE.md, is that intentional? I don't really know how templates work...

I'm not sure if requiring tests for every PR is a good idea, if someone doesn't feel comfortable writing tests then I'd rather have the feature without tests than no feature at all. Maybe we should either require tests or a `[WIP]` in the title if there are no tests so others can add them?

**Comment by LongLiveCHIEF** [comment]
I'm not sure why there are two PULL_REQUEST_TEMPLATE files, but it wasn't my intent. I think github might have added that, but I'll double check and update the PR.

I hear what you're saying, and I *do* think that it's important to make things easy for contributors, but at the same time... There are currently nearly 40,000 downloads of this plugin.

I think part of writing and contributing to open source software, is a responsibility to make sure that software will stand up without your direct involvement, since any contribution is likely to live on far beyond the use or life-cycle the submitter intended.

A test for a contribution is sort of like a representation of the contributor's interests. I see it as a way of programmatically consulting with all the prior contributors whenever a new contribution is added, because the test they wrote for their code is there to speak for them.

I also have been a maintainer of many other *large* projects, and if you don't require tests, then they wind up not being written, because maintainers of the project tend to have other things they spend their time focusing on, and they don't have time to write tests for every PR that comes in.

This means that after time, you wind up having a bunch of features with no test code, and it becomes more and more difficult to maintain the project as bugs creep in with each new PR... bugs that tests likely would have caught.

Then there's the jerkish/opinionated part of me that also adds, "a responsible programmer should be able to write tests", and allowing contributors to submit PR's without them isn't doing anything to help improve the community of developers who contribute to projects.

That same opinionated part of me won't agree to be a maintainer on a project that doesn't have this as a rule. I've worked with @lloeki before, and I sorta thought this was implied, based on projects we've collaborated on before. 

If you guys don't want to agree to this one, then I'll bow out. No hard feelings, but I have a metric butt-ton of other projects to maintain, and not having this rule makes it *just* easy enough to be a maintainer for this. I use it daily, but then again, it wouldn't be too difficult for me to just use vim instead.  Have to pick my battles... ya know?

**Comment by lloeki** [comment]
> I'm not sure if requiring tests for every PR is a good idea, if someone doesn't feel comfortable writing tests then I'd rather have the feature without tests than no feature at all.

The initial impetus of the project was to scratch a very obvious and visible itch of many folks. As such I said *screw it* and dived right in, to have some reasonable, working PoC. Then with a working solution, acquiring a critical mass features for the project to even matter was key. Doubled by the fact that I'm not a hotshot at JS testing, this led to a *features for sure, tests maybe* attitude (which is entirely my fault).

> I hear what you're saying, and I do think that it's important to make things easy for contributors, but at the same time... There are currently nearly 40,000 downloads of this plugin.

I think though that we have now a significant enough number of features that we can start to get ourselves out of that hole, and the [first law of holes is to stop digging](https://en.wikipedia.org/wiki/Law_of_holes). To that end, we should require tests before merge.

> Maybe we should either require tests or a [WIP] in the title if there are no tests so others can add them?

This is the part I am leaning towards. To allow contributions, tests should be required for a merge to be performed, but those tests may not need to be written by the PR author himself. A `needs tests` tag would be welcome to denote that.

@jazzpi, @LongLiveCHIEF, would that be OK for you?

**Comment by lloeki** [comment]
> I'm not sure why there are two PULL_REQUEST_TEMPLATE files, but it wasn't my intent. I think github might have added that, but I'll double check and update the PR.

Typical of a round-trip between case sensitive and case insensitive filesystems. You may have to `git mv` it to something else, maybe commit, and then go back to the correct case `.md`.

**Comment by LongLiveCHIEF** [comment]
> This is the part I am leaning towards. To allow contributions, tests should be required for a merge to be performed, but those tests may not need to be written by the PR author himself. A needs tests tag would be welcome to denote that.

Yes, that's a way of doing it that I hadn't considered, and I like it. I'm ok with that.

**Comment by LongLiveCHIEF** [comment]
@lloeki @jazzpi updated... look good to you?

**Comment by jazzpi** [comment]
Unfortunately only authorized users can add a label to an issue on GitHub, so I believe we should just require a `[needs tests]` in the title or something.

**Comment by lloeki** [comment]
Good point.

> we should just require a [needs tests] in the title or something.

Maybe have something in the template around "I have written tests for:" to make it even more obvious, then a collaborator can flag it with a label. Fine for me either way.

@LongLiveCHIEF nitpick: the guidelines wouldn't pass an INTERCAL compiler (excessively polite) ;-)

**Comment by LongLiveCHIEF** [comment]
> nitpick: the guidelines wouldn't pass an INTERCAL compiler (excessively polite) ;-)

You're the *only* person who has ever prompted me to reference the jargon file more than once. Congratulations.

Any suggestions?

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+3. The last line of your request should `/cc @jazzpi @LongLiveCHIEF`
+
+## Pull Requests
+
+- If the PR *fixes* or should result in the closure of any issues, use the `fixes #` or `closes #` syntax to ensure issue will
+close when your PR is merged
+- All pull-requests that fix a bug, or add a new feature, *must* have accompanying tests, before they will be merged. If you want
+to speed up the merge of your PR, please contribute these tests
+ - *note*: if you submit a PR, but are unsure how to write tests, please set the "needs tests" label
+- Please use the [pull request template](PULL_REQUEST_TEMPLATE.md) as a guide for submitting your PR.
```
Comment (lloeki) [inline .github/CONTRIBUTING.md]: Just remove this leading "please" and we're set.

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+1. Open a new issue, titled: `New Maintainer Request`
+2. Assign the issue to [@lloeki](https://github.com/lloeki)
+3. The last line of your request should `/cc @jazzpi @LongLiveCHIEF`
+
+## Pull Requests
+
+- If the PR *fixes* or should result in the closure of any issues, use the `fixes #` or `closes #` syntax to ensure issue will
+close when your PR is merged
+- All pull-requests that fix a bug, or add a new feature, *must* have accompanying tests, before they will be merged. If you want
```
Comment (lloeki) [inline .github/CONTRIBUTING.md]: Remove all those extraneous commas.

**Comment by LongLiveCHIEF** [comment]
@jazzpi @lloeki updated per both of your suggestions. Look good?

**Comment by lloeki** [comment]
All clear :+1:

**Comment by jazzpi** [comment]
Looks all fine, thanks! :+1:

**Pull request merged**